File: 알고리즘/이것이코딩테스트다스터디/다이나믹프로그래밍/합분해.py
#https://www.acmicpc.net/problem/2225
N , K = map(int, input().split())

dp = [[1 for _ in range(N+1) ] for _ in range(K)]

for i in range(1, K):
    for j in range(len(dp[i])):
        dp[i][j] = 0
        for k in range(j+1):
            dp[i][j] += dp[i-1][k]
            
        dp[i][j] = dp[i][j] % 1000000000
            
print(dp[K-1][N] % 1000000000)

File: Algorithm/C++/백준/다이나믹프로그래밍/합분해.cpp
// https://www.acmicpc.net/problem/2225

#include <iostream>
#include <vector>
#include <string>

using namespace std;

constexpr int DIV = 1000000000;

vector<vector<int>> dp;
int N, K;

int dfs(int value, int cnt)
{
    if (cnt > K || value > N)
    {
        return 0;
    }

    if (cnt == K && value == N)
    {
        return 1;
    }

    if (dp[value][cnt] != -1) return dp[value][cnt];

    int sum = 0;

    for (int n = 0; n <= N; n++)
    {
        sum += dfs(value + n, cnt + 1);
        sum %= DIV;
    }
    
    dp[value][cnt] = sum;
    return dp[value][cnt];
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    cin >> N >> K;

    dp.resize(N+1, vector<int>(K + 1, -1));

    cout << dfs(0, 0);
}
